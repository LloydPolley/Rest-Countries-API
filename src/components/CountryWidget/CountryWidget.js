import React from "react";
import { Link } from "react-router-dom";
import classNames from "classnames/bind";
import { formatNumber } from "../../data/rest";

import style from "./CountryWidget.module.scss";

const cx = classNames.bind(style);

const CountryWidget = ({ data }) => {
  const { name, flags, population, region, capital } = data;
  const { common } = name;

  return (
    <div
      data-testid="country-widget"
      key={common}
      className={cx("country", "element")}
    >
      <Link data-testid="country-link" to={`/${common}`}>
        <div
          className={cx("country__image")}
          style={{ backgroundImage: `url("${flags.svg}")` }}
        ></div>
      </Link>

      <div className={cx("country__data")}>
        <Link to={`/view?location=${common}`}>
          <h2 className="underline-swipe">{common}</h2>
        </Link>
        <p>
          Population: <span>{formatNumber(population)}</span>
        </p>
        <p>
          Region: <span>{region}</span>
        </p>
        <p>
          Capital: <span>{capital}</span>
        </p>
      </div>
    </div>
  );
};

export default CountryWidget;
