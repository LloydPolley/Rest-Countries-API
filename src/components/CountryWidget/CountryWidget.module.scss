@import "../../App.scss";

.country {
  overflow: hidden;
  border-radius: $borderRadius;
  z-index: 1;
  box-shadow: 0px 0px 10px 2px rgba(0, 0, 0, 0.001);

  &:hover {
    // border: 1px solid rgba(255, 255, 255, 1);
    box-shadow: 1px 1px 1px solid black;
  }

  a {
    text-decoration: none;
  }
  &__image {
    aspect-ratio: 16/9;
    object-fit: cover;
    background-position: center;
    background-repeat: no-repeat;
    overflow: hidden;
    background-size: cover;

    a {
      width: 100%;
    }
    img {
      width: 100%;
    }
  }
  &__data {
    padding: 30px;
    p {
      font-weight: bold;
    }
    span {
      font-weight: 400;
      margin-left: 5px;
    }
  }
}

@media (min-width: 700px) {
  .country-container {
    display: flex;
    flex-wrap: wrap;
  }
  .country {
    &__image {
      object-fit: contain;
      background-position: center;
    }
  }
}
