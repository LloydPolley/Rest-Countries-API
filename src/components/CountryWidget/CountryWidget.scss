.countryWidget{
    margin: 7% auto;
    overflow: hidden;
    border-radius: 10px;
    width: 80%;
    
    a{
        text-decoration: none;
    }
    &__image{
        width: 100%;
        display: flex;
        justify-content: center;
        overflow: hidden;
        img{
            width: 100%;
        }
    }
    &__data{
        padding: 1em;
    }
}

@media(min-width: 700px){
    .countryWidgetContainer{
        display: flex;
        flex-wrap: wrap;
    }
    .countryWidget{
        margin: 5% auto;
        width: 45%;

        &__image{
            height: 180px;
            object-fit: scale-down;

            img{
                width: 100%;
            }
        }
    }
}
@media(min-width: 1000px){
    .countryWidgetContainer{
        display: flex;
        flex-wrap: wrap;
    }
    .countryWidget{
        margin: 5% auto;
        width: 23%;
    }
}