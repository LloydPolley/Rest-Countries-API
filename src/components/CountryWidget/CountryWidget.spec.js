import React from "react";
import { render } from "@testing-library/react";
import { MemoryRouter } from "react-router-dom";
import CountryWidget from "./CountryWidget";

describe("CountryWidget", () => {
  const data = {
    name: { common: "Test Country" },
    flags: { svg: "test.svg" },
    population: 1000000,
    region: "Test Region",
    capital: "Test Capital",
  };

  it("renders correctly", () => {
    const { container } = render(
      <MemoryRouter>
        <CountryWidget data={data} />
      </MemoryRouter>
    );
    expect(container).toMatchSnapshot();
  });

  it("renders the correct data", () => {
    const { getByTestId, getByText } = render(
      <MemoryRouter>
        <CountryWidget data={data} />
      </MemoryRouter>
    );
    const countryWidgetElement = getByTestId("country-widget");
    const linkElement = getByTestId("country-link");
    const populationElement = getByText((content, element) => {
      // Custom text matcher function
      return element.tagName === "P" && content.startsWith("Population:");
    });
    const regionElement = getByText((content, element) => {
      // Custom text matcher function
      return element.tagName === "P" && content.startsWith("Region:");
    });
    const capitalElement = getByText((content, element) => {
      // Custom text matcher function
      return element.tagName === "P" && content.startsWith("Capital:");
    });

    expect(countryWidgetElement).toBeInTheDocument();
    expect(linkElement).toHaveAttribute("href", `/${data.name.common}`);
    expect(populationElement).toBeInTheDocument();
    expect(regionElement).toBeInTheDocument();
    expect(capitalElement).toBeInTheDocument();
  });

  it("renders the correct image URL", () => {
    const { getByTestId } = render(
      <MemoryRouter>
        <CountryWidget data={data} />
      </MemoryRouter>
    );
    const imageElement =
      getByTestId("country-widget").querySelector(".country__image");
    expect(imageElement).toHaveStyle(
      `background-image: url("${data.flags.svg}")`
    );
  });
});
