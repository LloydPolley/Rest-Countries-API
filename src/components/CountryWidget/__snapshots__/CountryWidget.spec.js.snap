// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CountryWidget renders correctly 1`] = `
<div>
  <div
    class="country element"
    data-testid="country-widget"
  >
    <a
      data-testid="country-link"
      href="/Test Country"
    >
      <div
        class="country__image"
        style="background-image: url(test.svg);"
      />
    </a>
    <div
      class="country__data"
    >
      <a
        href="/Test Country"
      >
        <h2
          class="underline-swipe"
        >
          Test Country
        </h2>
      </a>
      <p>
        Population: 
        <span>
          1,000,000
        </span>
      </p>
      <p>
        Region: 
        <span>
          Test Region
        </span>
      </p>
      <p>
        Capital: 
        <span>
          Test Capital
        </span>
      </p>
    </div>
  </div>
</div>
`;
